src/guardian/staking-module-guard/staking-module-guard.service.ts

```diff
@@ -111,6 +111,100 @@ export class StakingModuleGuardService {
     );
   }
 
+  isEarlyDeposit(
+    firstEvent: VerifiedDepositEvent,
+    secondEvent: VerifiedDepositEvent,
+  ) {
+    const isSameBlock = firstEvent?.blockNumber === secondEvent.blockNumber;
+
+    let isEarlyDeposit = false;
+
+    if (isSameBlock) {
+      isEarlyDeposit = firstEvent?.logIndex < secondEvent.logIndex;
+    } else {
+      isEarlyDeposit = firstEvent?.blockNumber < secondEvent.blockNumber;
+    }
+
+    return isEarlyDeposit;
+  }
+  /**
+   * Method is not taking into account WC rotation since historical deposits was checked manually
+   * @param blockData
+   * @returns
+   */
+  async getHistoricalFrontRun(blockData: BlockData) {
+    const { depositedEvents, lidoWC } = blockData;
+    const potentialLidoDepositsEvents = depositedEvents.events.filter(
+      ({ wc, valid }) => wc === lidoWC && valid,
+    );
+
+    this.logger.log('potential lido deposits events count', {
+      count: potentialLidoDepositsEvents.length,
+    });
+
+    const potentialLidoDepositsKeysMap: Record<string, VerifiedDepositEvent> =
+      {};
+
+    potentialLidoDepositsEvents.forEach((event) => {
+      if (potentialLidoDepositsKeysMap[event.pubkey]) {
+        const existed = potentialLidoDepositsKeysMap[event.pubkey];
+        const isExisted = this.isEarlyDeposit(existed, event);
+        // this should not happen, since Lido deposits once per key.
+        // but someone can still make such a deposit.
+        if (isExisted) return;
+      }
+      potentialLidoDepositsKeysMap[event.pubkey] = event;
+    });
+
+    const duplicatedDepositEvents: VerifiedDepositEvent[] = [];
+
+    depositedEvents.events.forEach((event) => {
+      if (potentialLidoDepositsKeysMap[event.pubkey] && event.wc !== lidoWC) {
+        duplicatedDepositEvents.push(event);
+      }
+    });
+
+    this.logger.log('duplicated deposit events', {
+      count: duplicatedDepositEvents.length,
+    });
+
+
+    const frontRunnedDepositEvents = duplicatedDepositEvents.filter(
+      (suspectedEvent) => {
+        // get event from lido map
+        const sameKeyLidoDeposit =
+          potentialLidoDepositsKeysMap[suspectedEvent.pubkey];
+
+        if (!sameKeyLidoDeposit) throw new Error('expected event not found');
+
+        return this.isEarlyDeposit(suspectedEvent, sameKeyLidoDeposit);
```

Comment:
there is an attack still:

> invalid deposit with Lido wc + pubkey ⇒ frontrunned deposit with malicious wc + Lido pubkey ⇒ valid deposit with Lido wc + pubkey

Reply:
bc here the invalid event might be stored as a reference:
![image](https://github.com/lidofinance/lido-council-daemon/assets/961317/5430112f-8624-431c-b797-abe6acd938a3)

Reply:
fixed